File: pages/refactorings/extract-method/03-expression-eval/Spec.scala
//> using scala 3.3.4
//> using dep qa.hedgehog::hedgehog-core:0.14.0
//> using dep qa.hedgehog::hedgehog-runner:0.14.0
import hedgehog.*, hedgehog.core.*, hedgehog.runner.*

object Props extends Properties:
  def tests: List[Test] = List(
    property("eval: Before == After", evalAgrees),
    property("dividing by zero has no value, never throws", divByZeroIsNone),
  )

  import Before.Expr, Before.Expr.*

  val genLit: Gen[Expr] = Gen.int(Range.linear(-20, 20)).map(Lit(_))
  def genExpr(depth: Int): Gen[Expr] =
    if depth == 0 then genLit
    else
      val sub = genExpr(depth - 1)
      Gen.choice1(
        genLit,
        for l <- sub; r <- sub yield Add(l, r),
        for l <- sub; r <- sub yield Mul(l, r),
        for l <- sub; r <- sub yield Div(l, r),
      )

  def toAfter(e: Expr): After.Expr = e match
    case Lit(n)    => After.Expr.Lit(n)
    case Add(l, r) => After.Expr.Add(toAfter(l), toAfter(r))
    case Mul(l, r) => After.Expr.Mul(toAfter(l), toAfter(r))
    case Div(l, r) => After.Expr.Div(toAfter(l), toAfter(r))

  def evalAgrees: Property =
    for e <- genExpr(4).forAll
    yield Before.eval(e) ==== After.eval(toAfter(e))

  def divByZeroIsNone: Property =
    for e <- genExpr(3).forAll
    yield Before.eval(Div(e, Lit(0))) ==== None and After.eval(toAfter(Div(e, Lit(0)))) ==== None

@main def spec(): Unit =
  val results = Props.tests.map { t =>
    val r = Property.check(t.withConfig(PropertyConfig.default), t.result, Seed.fromTime())
    println(Test.renderReport("Props", t, r, ansiCodesSupported = false))
    r.status
  }
  if !results.forall(_ == Status.ok) then sys.exit(1)

File: pages/refactorings/extract-method/03-expression-eval/Spec.hs
{-# LANGUAGE OverloadedStrings #-}
module Main where

import Control.Monad (unless)
import System.Exit (exitFailure)
import Hedgehog
import qualified Hedgehog.Gen as Gen
import qualified Hedgehog.Range as Range
import qualified Before
import qualified After

genExpr :: Gen Before.Expr
genExpr = Gen.recursive Gen.choice
  [ Before.Lit <$> Gen.int (Range.linear (-20) 20) ]
  [ Gen.subterm2 genExpr genExpr Before.Add
  , Gen.subterm2 genExpr genExpr Before.Mul
  , Gen.subterm2 genExpr genExpr Before.Div
  ]

toAfter :: Before.Expr -> After.Expr
toAfter (Before.Lit n)   = After.Lit n
toAfter (Before.Add l r) = After.Add (toAfter l) (toAfter r)
toAfter (Before.Mul l r) = After.Mul (toAfter l) (toAfter r)
toAfter (Before.Div l r) = After.Div (toAfter l) (toAfter r)

prop_eval_agrees :: Property
prop_eval_agrees = property $ do
  e <- forAll genExpr
  Before.eval e === After.eval (toAfter e)

-- A left operand without a value short-circuits: the right operand is never
-- forced, before and after. (Division by zero has no value, so it does not throw here.)
prop_short_circuit_preserved :: Property
prop_short_circuit_preserved = property $ do
  e <- forAll genExpr
  Before.eval (Before.Add (Before.Div e (Before.Lit 0)) undefined) === Nothing
  After.eval (After.Add (After.Div (toAfter e) (After.Lit 0)) undefined) === Nothing

main :: IO ()
main = do
  ok <- checkParallel $ Group "Props"
    [ ("eval: Before == After", prop_eval_agrees)
    , ("short-circuit on a valueless left operand is preserved", prop_short_circuit_preserved)
    ]
  unless ok exitFailure
